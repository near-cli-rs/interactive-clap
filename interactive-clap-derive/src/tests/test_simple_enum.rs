use super::pretty_codegen;

#[test]
fn test_simple_enum() {
    let input = syn::parse_quote! {
        pub enum Mode {
            /// Prepare and, optionally, submit a new transaction with online mode
            Network,
            /// Prepare and, optionally, submit a new transaction with offline mode
            Offline,
        }
    };

    let interactive_clap_codegen = crate::derives::interactive_clap::impl_interactive_clap(&input);
    insta::assert_snapshot!(pretty_codegen(&interactive_clap_codegen));

    let step_one_output = syn::parse_quote! {
        pub enum CliMode {
            /// Prepare and, optionally, submit a new transaction with online mode
            Network,
            /// Prepare and, optionally, submit a new transaction with offline mode
            Offline,
        }
    };

    let to_cli_args_codegen = crate::derives::to_cli_args::impl_to_cli_args(&step_one_output);
    insta::assert_snapshot!(pretty_codegen(&to_cli_args_codegen));
}

#[test]
fn test_simple_enum_with_strum_discriminants() {
    let input = syn::parse_quote! {
        #[strum_discriminants(derive(EnumMessage, EnumIter))]
        /// A little beautiful comment about our choice
        pub enum Mode {
            /// Prepare and, optionally, submit a new transaction with online mode
            #[strum_discriminants(strum(message = "Yes, I keep it simple"))]
            Network,
            /// Prepare and, optionally, submit a new transaction with offline mode
            #[strum_discriminants(strum(
                message = "No, I want to work in no-network (air-gapped) environment"
            ))]
            Offline,
        }
    };

    let interactive_clap_codegen = crate::derives::interactive_clap::impl_interactive_clap(&input);
    insta::assert_snapshot!(pretty_codegen(&interactive_clap_codegen));

    let step_one_output = syn::parse_quote! {
        pub enum CliMode {
            /// Prepare and, optionally, submit a new transaction with online mode
            Network,
            /// Prepare and, optionally, submit a new transaction with offline mode
            Offline,
        }
    };

    let to_cli_args_codegen = crate::derives::to_cli_args::impl_to_cli_args(&step_one_output);
    insta::assert_snapshot!(pretty_codegen(&to_cli_args_codegen));
}
