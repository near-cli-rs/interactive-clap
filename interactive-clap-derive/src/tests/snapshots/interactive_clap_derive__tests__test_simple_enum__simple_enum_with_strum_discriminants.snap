---
source: interactive-clap-derive/src/tests/test_simple_enum.rs
expression: pretty_codegen(&interactive_clap_codegen)
---
#[derive(Debug, Clone, clap::Parser, interactive_clap::ToCliArgs)]
pub enum CliMode {
    /// Prepare and, optionally, submit a new transaction with online mode
    Network,
    /// Prepare and, optionally, submit a new transaction with offline mode
    Offline,
}
impl interactive_clap::ToCli for Mode {
    type CliVariant = CliMode;
}
pub type InteractiveClapContextScopeForMode = ModeDiscriminants;
impl interactive_clap::ToInteractiveClapContextScope for Mode {
    type InteractiveClapContextScope = InteractiveClapContextScopeForMode;
}
impl From<Mode> for CliMode {
    fn from(command: Mode) -> Self {
        match command {
            Mode::Network => Self::Network,
            Mode::Offline => Self::Offline,
        }
    }
}
impl interactive_clap::FromCli for Mode {
    type FromCliContext = ();
    type FromCliError = color_eyre::eyre::Error;
    fn from_cli(
        mut optional_clap_variant: Option<<Self as interactive_clap::ToCli>::CliVariant>,
        context: Self::FromCliContext,
    ) -> interactive_clap::ResultFromCli<
        <Self as interactive_clap::ToCli>::CliVariant,
        Self::FromCliError,
    >
    where
        Self: Sized + interactive_clap::ToCli,
    {
        loop {
            return match optional_clap_variant {
                Some(CliMode::Network) => {
                    interactive_clap::ResultFromCli::Ok(CliMode::Network)
                }
                Some(CliMode::Offline) => {
                    interactive_clap::ResultFromCli::Ok(CliMode::Offline)
                }
                None => {
                    match Self::choose_variant(context.clone()) {
                        interactive_clap::ResultFromCli::Ok(cli_args) => {
                            optional_clap_variant = Some(cli_args);
                            continue;
                        }
                        result => return result,
                    }
                }
            };
        }
    }
}
impl Mode {
    pub fn choose_variant(
        context: (),
    ) -> interactive_clap::ResultFromCli<
        <Self as interactive_clap::ToCli>::CliVariant,
        <Self as interactive_clap::FromCli>::FromCliError,
    > {
        use interactive_clap::SelectVariantOrBack;
        use inquire::Select;
        use strum::{EnumMessage, IntoEnumIterator};
        let selected_variant = Select::new(
                concat!(r" A little beautiful comment about our choice",).trim(),
                ModeDiscriminants::iter()
                    .map(SelectVariantOrBack::Variant)
                    .chain([SelectVariantOrBack::Back])
                    .collect(),
            )
            .prompt();
        match selected_variant {
            Ok(SelectVariantOrBack::Variant(variant)) => {
                let cli_args = match variant {
                    ModeDiscriminants::Network => CliMode::Network,
                    ModeDiscriminants::Offline => CliMode::Offline,
                };
                return interactive_clap::ResultFromCli::Ok(cli_args);
            }
            Ok(SelectVariantOrBack::Back) => return interactive_clap::ResultFromCli::Back,
            Err(
                inquire::error::InquireError::OperationCanceled
                | inquire::error::InquireError::OperationInterrupted,
            ) => return interactive_clap::ResultFromCli::Cancel(None),
            Err(err) => return interactive_clap::ResultFromCli::Err(None, err.into()),
        }
    }
    pub fn try_parse() -> Result<CliMode, clap::Error> {
        <CliMode as clap::Parser>::try_parse()
    }
    pub fn parse() -> CliMode {
        <CliMode as clap::Parser>::parse()
    }
    pub fn try_parse_from<I, T>(itr: I) -> Result<CliMode, clap::Error>
    where
        I: ::std::iter::IntoIterator<Item = T>,
        T: ::std::convert::Into<::std::ffi::OsString> + ::std::clone::Clone,
    {
        <CliMode as clap::Parser>::try_parse_from(itr)
    }
}
